### DISCUSSION TITLE
Alternative to new, mutable libraries operating on mutable records containing pure data: add syntactic sugar to simplify updates of mutable locations holding pure contents:

### DISCUSSION THREAD

**Discussion by crusso** [caffeinelabs/motoko-core#58] [Ideas]
Silly idea perhaps, but if we are serious about using the pure data structures to implement stateful ones, maybe, instead of providing wrapper classes holding a mutable cell, it would be enough to just add some sugar for updating a mutable location passed to a function call. Something like adding an _ <- _ operator (read "becomes"):
```
let NatMap = OrderedMap<Nat>(Nat.compare);
var map : Map<Nat, Text> = NatMap.empty<Text>();
map := NatMap.add(map, 1, "one");
map <- NatMap.add(1, "one"); // sugar for previous line

```
Sugar:

```
<exp0> <- <exp><inst>(<exp1>, ..., <expn>) :=
   <exp0> := <exp><inst>(<exp0>, <exp1>, ..., <expn>)
```

(perhaps with the restriction that exp0 is an lvar to avoid code duplication).

Probably a bad idea... (edited)

**Comment by rvanasa** [comment]
Something along the lines of this idea could be nice. Here's a variation which would preserve the tuple structure and give more flexibility for the position of the modified value:

```motoko
NatMap.add(-> map, 1, "one");

NatMap.add(ref map, 1, "one");

NatMap.add(*map, 1, "one");
```

A downside is that this might be hard to understand for beginners coming from other languages, but I like this general concept as a way to simplify using purely functional data structures.

It's also worth mentioning that `Vec` does not have a purely functional equivalent, which means that the syntax might be inconsistent between data structures. This might be okay but is perhaps a reason to keep two separate APIs.
